### ISSUE TITLE
Xpetra::CrsMatrix::doImport() fails with Negative count into MPI_Send

### ISSUE THREAD

**Issue by SG38** [trilinos/Trilinos#2522]
The MPI_Send() has a Negative count when importing big data set of a matrix ; same issue with MultiVector and CrsMatrix. When distributing the matrix with an import of the SerialMatrix rows:

_RCP<Xpetra::Import<SOL_LO, SOL_GO, SOL_EPETRA_TYPE> > importer = Xpetra::ImportFactory<SOL_LO, SOL_GO, SOL_EPETRA_TYPE>::Build(SerialMap, DistributedMap);_
_DistributedMatrix->doImport(*SerialMatrix, *importer, Xpetra::INSERT);_

It throws:

> Fatal error in MPI_Send: Invalid count, error stack:
> MPI_Send(201): MPI_Send(buf=0x7ffb33824100, count=-2131745804, MPI_CHAR, dest=1, tag=1, MPI_COMM_WORLD) failed
> MPI_Send(117): Negative count, value is -2131745804

Configuration is:

(1) Trilinos 12.12.1
Linux 64 bits

(2) CMakeCache:

- Tpetra_INST_INT_LONG_LONG:BOOL=ON
- HAVE_TEUCHOS_LONG_LONG_INT:INTERNAL=ON
- HAVE_TEUCHOS___INT64:INTERNAL=1

(3) program:
- Xpetra::UseTpetra
- Xpetra::CrsMatrix<int, signed long long, KokkosClassic::DefaultNode::DefaultNodeType>
- Teuchos::Comm< int > comm;

@trilinos/Teuchos

To be honest, I do not understand the source code of the Teuchos communicator. More precisely, i do not find any source code where the _send_ counter would be a long long. Is there a way to distribute big matrices?

Thank you in advance,

**Comment by aprokop** [comment]
@trilinos/teuchos @trilinos/xpetra 

This is clearly an integer overflow. And particularly weird to me is the presence of `MPI_CHAR`. I thought we got rid of that?

**Comment by ibaned** [comment]
Note that MPI itself defines counts as being of type `int`, so if you want to send more than `INT_MAX` of something you need to break it up into smaller messages. Maybe somewhere between Xpetra, Tpetra, and Teuchos there should be additional logic for this.

**Comment by mhoemmen** [comment]
I assert that this is a limitation of MPI implementations itself, because MPI functions take counts as `int`.  Jeff Hammond wrote an MPI implementation that supports 64-bit counts:

https://github.com/jeffhammond/BigMPI

If we want to support 64-bit counts, then that would be the library to use.  I would prefer not to clutter up Trilinos with this logic.

`Epetra_DistObject` turns a lot of packet types into `char`, hence `MPI_CHAR`.  In particular, CrsMatrix Import/Export needs to send both integers and floating-point values.  CrsMatrix packs them together into a `char` buffer.

`Teuchos::Comm` just wraps MPI.  There's nothing particularly mysterious about it.  The wrapper only gets tricky for types that require serialization.

**Comment by mhoemmen** [comment]
@ibaned @aprokop @trilinos/xpetra 

I am tempted to mark this as "wontfix," unless someone really has a plan and funding to plug in a solution like Jeff's.  It's not quite plug and play, since the interfaces may differ.

**Comment by aprokop** [comment]
@mhoemmen Should not there simply be a check for the situation when the number of counts exceeds MAX_INT somewhere in Tpetra/Teuchos?

**Comment by mhoemmen** [comment]
@aprokop That could be helpful, but I think of that as a separate issue.

I was kinda hoping not to be Sole Keeper of the Mystical Teuchos Wrappers For All Time ;-)

**Comment by nmhamster** [comment]
@mhoemmen -

> Epetra_DistObject turns a lot of packet types into char, hence MPI_CHAR. In particular, CrsMatrix Import/Export needs to send both integers and floating-point values. CrsMatrix packs them together into a char buffer.

Is it not possible to use the smaller of the widths for the integer and floating-point values as the base rather than `MPI_CHAR`? This would increase the number of items we could send.

**Comment by mhoemmen** [comment]
@nmhamster That's possible, but touching Epetra's communication layer historically has a high risk of breaking something.  Such a design would also need care to avoid violating C++ aliasing rules.

The general problem right now with the communication layer is staffing.

**Comment by SG38** [comment]
Hi, 
I see that's causing a heated discussion. I agree with @aprokop  to check that it does not exceed MAX_INT should be a robust solution. Probably increasing MPI_CHAR to MPI_UNSIGNED_LONG...
However, it is not implemented in Teuchos yet. Then, my question is : "how do you solve big systems (> 100M equations) at the moment? 
Maybe it is not a fix to do; maybe I do not the right thing?
Precisely because (1) I read the whole equations from a file, (2) I fillin a sparse matrix and then (3) I distribute it through the import or export procedure of the sparse matrix. Same thing for the MultiVector matrices of RHS and LHS, that can be very big also. 
Is there a better way to solve A.X = B reading very big matrices in parallel? I have seen that you can treat solves with billions of equations.

Thank you for your patience,

**Comment by ibaned** [comment]
@SG38 maybe try to form the sparse matrix in parallel from the start? The problem is having the entire sparse matrix on a single MPI rank.

**Comment by mhoemmen** [comment]
@SG38 wrote:
> "how do you solve big systems (> 100M equations) at the moment?

What @ibaned said.  Everything starts and stays distributed: the mesh, the graph, the matrix, the right-hand side vector, and the solution vector.  We've solved problems big enough not to fit in `int`.

**Comment by SG38** [comment]
Would you mind to guide me in this direction please? I need an example please. I understand that it is a better way to avoid to have a big matrix in memory. The graph must be distributed early in the flow.
Thank you

**Comment by mhoemmen** [comment]
@SG38 Are you working with an application that has a distributed mesh?

**Comment by SG38** [comment]
@mhoemmen

**Comment by mhoemmen** [comment]
@SG38 I'm not an expert on distributed meshes.  You could use Sandia's open-source library STK (the Sierra Tool Kit) for that purpose.  The STK manual could be a good start.

**Comment by kddevin** [comment]
MPI is limited to have integers as its send/receive size in the MPI API.  In this case, you can likely avoid the problem by not building the entire matrix on rank 0 and then redistributing.  Instead, assign each processor a set of rows, read the matrix in chunks, broadcast each chunk, and let each processor store the entries corresponding to its rows.  Roughly, 

```
create default Tpetra Map with nGlobalRows over nProcessors 
myFirstRow = map.getMinGlobalIndex() 
myLastRow = map.getMaxGlobalIndex() 
nChunks = nNonZeros / chunkSize 
for each chunk in nChunks
    if (myRank == 0)  read chunk
    broadcast chunk
    for each nonzero (i,j) in chunk
         if (i >= myFirstRow && i <= myLastRow)  store (i,j) in my processor
```
Then you can redistribute later to account for better balance and locality (e.g., by using Zoltan2), but that redistribution will be more balanced with respect to sends/receives.
